### DISCUSSION TITLE
API 테스트에서 응답 가져오는 방법

### DISCUSSION THREAD

**Discussion by jseop-lim** [8percent/styleguide#64] [Ideas]
JSON을 반환하는 API 테스트를 할 때 rest_framework.test.APIClient 혹은 django.test.Client 객체(이하 client 객체)를 활용하여 응답 결과를 비교합니다.

client의 get()/post()/put()/patch()/delete() 같은 메서드의 반환 값을 response라 할 때, response에서 API 응답 JSON을 가져오는 방법이 두 가지 있습니다.
* `response.data`
* `response.json()`

아래는 `/my-api/`라는 가상의 API에 대한 테스트 예제입니다.
```python
from datetime import datetime
from django.utils import timezone
from rest_framework.test import APIClient


def test_get_200():
    client = APIClient()

    response = client.get(path="/my-api/")

    assert response.status_code == 200
    # data를 비교
    assert response.data == {
        "token": "test",
        "token_timestamp": timezone.make_aware(datetime(2023, 3, 3, 0, 0, 0)),
    }
    # json()을 비교
    assert response.json() == {
        "token": "test",
        "tokenTimestamp": "2023-03-03T00:00:00+09:00",
    }
```

위 예제에서처럼 `.data`와 `.json()`에는 두 가지 차이가 있습니다.
* `.json()`에는 settings에 설정된 CamelCaseJSONRenderer가 적용되나 `.data`는 아님
* `.json()`은 직렬화된 결과가 담기나 `.data`에는 직렬화되기 전 파이썬 객체가 담김

직렬화 및 Renderer의 테스트 범위 포함 여부에 따라 적절한 방법을 선택하면 좋을 것 같습니다.

**Comment by jseop-lim** [comment]
.json()이 실제 클라이언트가 마주하는 응답값과 유사하므로 .json()을 사용하는 것은 어떨까요?

**Reply by cbscsm** [reply]
동의합니다.
API 테스트는 API 호출자의 관점에서 작성되어야 한다고 생각하기 때문입니다. 만약 객체로 검사해야하는 경우가 있다면 API 테스트가 아닌 다른 단계(예를들면 서비스)에서 이뤄지는 게 맞다고 생각합니다.

**Comment by jseop-lim** [comment]
### API 테스트가 end to end test인가?
서비스를 모킹, 입력과 출력부만 테스트
serializer나 service는 중복 최소화를 위해 각자 범위에서 테스트
or 테스트의 중복 허용 가능
url 테스트 필요? - 테스트 위치 찾기 - 디렉토리 위치를 잘 챙기자

### client 생성 방법
pyest client (객체 django client와 동일)
django client
drf apiclient

### API 테스트 필수 사항
* 서비스는 이미 테스트가 되어 있다. 서비스는 모킹
* url 호출, 응답을 테스트
* 서비스가 실패했을 때 응답
* 파라미터가 많은 경우 serializer 테스트로 분리

**Comment by jseop-lim** [comment]
테스트 코드 가독성을 좋게 만들면 테스트에 관심을 가지게 되지 않을까?

**Comment by jseop-lim** [comment]
Factory를 보완하여 자주 사용되는 객체를 한 방에 생성해주는 기능을 제공하자.
하지만 경우별 Factory를 미리 만들어두면 있는데도 몰라서 안쓰는 경우가 있다.
